compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -3232,6 +3232,12 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
   end provablyDisjointClasses
 
   private def provablyDisjointTypeArgs(cls: ClassSymbol, args1: List[Type], args2: List[Type], pending: util.HashSet[(Type, Type)])(using Context): Boolean =
+    // sjrd: I will not be surprised when this causes further issues in the future.
+    // This is a compromise to be able to fix #21295 without breaking the world.
+    def cannotBeNothing(tp: Type): Boolean = tp match
+      case tp: TypeParamRef => cannotBeNothing(tp.binder.paramInfos(tp.paramNum))
```

Comment:
Btw, I recently added `.paramInfo` to TypeParamRef, as in `cannotBeNothing(tp.paramInfo)`.